std/bigint.d

```diff
@@ -49,44 +50,113 @@ private:
     bool sign = false;
 public:
     /**
-     * Construct a BigInt from a decimal or hexadecimal string.
+     * Construct a BigInt from a decimal or hexadecimal string. The number must
+     * be in the form of a decimal or hex literal. It may have a leading $(D +)
+     * or $(D -) sign, followed by "0x" or "0X" if hexadecimal. Underscores are
+     * permitted in any location.
      *
-     * The number must be in the form of a D decimal or hex literal. It may
-     * have a leading + or - sign; followed by "0x" if hexadecimal.
-     * Underscores are permitted.  An empty string is treated as "0".
+     * Params:
+     *     s = a bidirectional range of any character type
      *
-     * Throws: ConvException if invalid character found, or string is empty.
+     * Throws:
+     *     $(D ConvException) if the string doesn't represent a valid number
      */
-    this(T : const(char)[])(T s, string file = __FILE__, size_t line = __LINE__)
-        pure
+    this(Range)(Range s) if (
+        isBidirectionalRange!Range &&
+        isSomeChar!(ElementType!Range) &&
+        !isInfinite!Range)
     {
-        if (s.length == 0)
-            throw new ConvException("Can't initialize BigInt with "~
-                                    "empty string", file, line);
+        import std.range : chain, walkLength;
+        import std.algorithm.iteration : filter, filterBidirectional;
+        import std.algorithm.comparison: equal;
+        import std.exception : enforce;
+        import std.conv : ConvException;
+
+        enforce!ConvException(!s.empty, "Can't initialize BigInt with an empty range");
+
+        ElementType!Range[2] buff;
         bool neg = false;
-        if (s[0] == '-') {
-            neg = true;
-            s = s[1..$];
-        } else if (s[0] == '+') {
-            s = s[1..$];
-        }
-        data = 0UL;
         bool ok;
-        assert(isZero());
-        if (s.length > 2 && (s[0..2] == "0x" || s[0..2] == "0X"))
+        auto filtered_range = s.filterBidirectional!(a => a != '_');
+        auto len = filtered_range.save.walkLength;
+        ubyte i;
+
+        data = 0UL;
+
+        // check for signs and if the string is a hex value
+        if (len > 1)
+        {
+            for (i = 0; i < 2; i++)
```

Comment:
I find this logic needlessly convoluted. Sign checking doesn't need to be interwoven with checking for `0x`. It should be a very simple, 2-step process:
````
// Check sign
if (range.front == '+')
    range.popFront(); // skip '+'
else if (range.front == '-')
{
    neg = true;
    range.popFront();
}

// Check for hex prefix
if (len > 2 && range.save.startsWith("0x"))
    // is hex
else
   // is decimal
````

Furthermore, filtering out `_` seems to be done prematurely. Do we really support literals of the form `_-_0_x_1234_567A`?? I would think that the negative sign and the `0x` prefix should not allow any intervening `_`'s. So really, sign checking and hex prefix checking should be done first, and *then* `_` gets filtered out.